# Krikey's Dataset - Data Analysis
In this notebook, we use the dataset provided by the krikey team and first perform descriptive data analysis to understand the data better. Futher, we go about with exploratory data analysis. Every part of the analysis is initiated by posing a question that is deemed to be useful for the business. All along this notebook, clear description and visualizations are provided as and when required. Table below is organization of the work carried out in this notebook as part of the data analysis. 

## Table of Contents
1. Import Python libraries
2. User-Defined Helper Functions
3. Load data files
4. Descriptive data analysis
5. Exploratory data analysis

***

## Import Python libraries
In this section, we import libraries as per requirement and check if the libraries are loading correctly by printing out the version of the library that is loaded.

In [286]:
#!pip install geopy
#!pip install folium

In [379]:
# Manage files/folders
import os

# Data processing analysis library
import numpy as np
import pandas as pd 
import calendar

# Pandas profiling library - A faster, effecient way in profiling the data
from pandas_profiling import ProfileReport

# Data viz library
import matplotlib.pyplot as plt
import matplotlib as mpl
import matplotlib.cm as cm
import seaborn as sns
import pandas_bokeh
import folium
from folium.plugins import HeatMap, MarkerCluster

# Geopy module
from geopy.geocoders import Nominatim

# Setting default viz options for pandas 
pd.set_option('plotting.backend', 'pandas_bokeh')

In [8]:
# check library versions
print(f'>>> Pandas version:{pd.__version__}')
print(f'>>> Seaborn version:{sns.__version__}')

# check pandas_bokeh
pandas_bokeh.output_notebook()

>>> Pandas version:1.4.1
>>> Seaborn version:0.11.2


Loading BokehJS ...

***

## User-Defined Helper Functions
This section is a collection of user-defined helper functions that are used all along this notebook. A better way to showcase these user-defined functions is to wrap them in a python files and import functions from the file. For this challenge, we show all the function in this notebook itself. 

In [302]:
def get_address_from_lat_lng(lat, lng):
    """
        Returns a Country for a set of lat, lng (inputs) 
        - using GeoPy Nominatm python library
    """
    geolocator = Nominatim(user_agent="krikey_challenge")
    # Location
    location = geolocator.reverse([str(lat), str(lng)])

    # Country
    country = location.raw['address']['country']
    
    return country


def plot_dots_folium(point):
    """
        Returns set of circle markers on top of base folium map 
    """
    folium.CircleMarker(location=[point.lat,point.lng],
                       radius=5,
                       weight=2,
                       popup=[point.lat, point.lng, point.id],
                       fill_color='#000000').add_to(m)

***

## Load Data & Check for data quality issues
In this section, we load the data from the set of csv files provided by krikey team. Further, we check the data for any data quality issues. These preliminary data quality checks will help us document any assumptions (or limitations concerning the data) before starting with the data analysis. This step will assist us in developing sensible business questions in the later stages. Our understanding of data in this section is documented in the **Notes on Data and Quality** sub-section.

### Load Data
Data is provided in 4 different csv files. Each csv file has specific information as below:

1. **watchedVideo**: This contains all occurrences of the watchedVideo event.
2. **users**: This table contains the registered users
3. **videos**: This table contains all created videos
4. **feeds**: This contains the feeds the videos are served on

In [49]:
# Set PATH for data files
PATH_DATA    = os.path.join(os.getcwd(),'data')
FILE_USERS   = os.path.join(PATH_DATA,'users.csv')
FILE_VIDEOS  = os.path.join(PATH_DATA,'videos.csv')
FILE_FEEDS   = os.path.join(PATH_DATA,'feeds.csv')
FILE_WATCHED = os.path.join(PATH_DATA,'watchedVideo.csv')

print('|------------------------------------------------------------------|')
print(f' PATH to DATA folder: {PATH_DATA}')
print(f'     - PATH to csv file: {FILE_USERS}')
print(f'     - PATH to csv file: {FILE_VIDEOS}')
print(f'     - PATH to csv file: {FILE_FEEDS}')
print(f'     - PATH to csv file: {FILE_WATCHED}')
print('|------------------------------------------------------------------|')

|------------------------------------------------------------------|
 PATH to DATA folder: /home/jovyan/work/data
     - PATH to csv file: /home/jovyan/work/data/users.csv
     - PATH to csv file: /home/jovyan/work/data/videos.csv
     - PATH to csv file: /home/jovyan/work/data/feeds.csv
     - PATH to csv file: /home/jovyan/work/data/watchedVideo.csv
|------------------------------------------------------------------|


In [153]:
# Load "users" data file
df_users = pd.read_csv(FILE_USERS, index_col=0)
print(f' >>> Loaded: {FILE_USERS} file has {df_users.shape[0]} rows and {df_users.shape[1]} columns')

# Load "videos" data file
df_videos = pd.read_csv(FILE_VIDEOS, index_col=0)
print(f' >>> Loaded: {FILE_VIDEOS} file has {df_videos.shape[0]} rows and {df_videos.shape[1]} columns')

# Load "feeds" data file
df_feeds = pd.read_csv(FILE_FEEDS, index_col=0)
print(f' >>> Loaded: {FILE_FEEDS} file has {df_feeds.shape[0]} rows and {df_feeds.shape[1]} columns')

# Load "watchedvideos" data file
df_watched = pd.read_csv(FILE_WATCHED, index_col=0)
print(f' >>> Loaded: {FILE_WATCHED} file has {df_watched.shape[0]} rows and {df_watched.shape[1]} columns')

 >>> Loaded: /home/jovyan/work/data/users.csv file has 1000 rows and 4 columns
 >>> Loaded: /home/jovyan/work/data/videos.csv file has 5000 rows and 10 columns
 >>> Loaded: /home/jovyan/work/data/feeds.csv file has 10 rows and 2 columns
 >>> Loaded: /home/jovyan/work/data/watchedVideo.csv file has 50000 rows and 8 columns


### Check Data Quality

In [154]:
# Check - missing values in users, videos, feeds, watchedVideo
print(f' >> No. of Missing Values in users.csv: {df_users.isnull().sum().sum()}')
print(f' >> No. of Missing Values in videos.csv: {df_videos.isnull().sum().sum()}')
print(f' >> No. of Missing Values in feeds.csv: {df_feeds.isnull().sum().sum()}')
print(f' >> No. of Missing Values in watchedVideo.csv: {df_watched.isnull().sum().sum()}')

 >> No. of Missing Values in users.csv: 0
 >> No. of Missing Values in videos.csv: 0
 >> No. of Missing Values in feeds.csv: 0
 >> No. of Missing Values in watchedVideo.csv: 0


In [155]:
# Check - Duplicate Rows in users, videos, feeds, watchedVideo
print(f' >> No. of Duplicate Rows in users.csv: {df_users.duplicated().sum()}')
print(f' >> No. of Duplicate Rows in videos.csv: {df_videos.duplicated().sum()}')
print(f' >> No. of Duplicate Rows in feeds.csv: {df_feeds.duplicated().sum()}')
print(f' >> No. of Duplicate Rows in watchedVideo.csv: {df_watched.duplicated().sum()}')

 >> No. of Duplicate Rows in users.csv: 0
 >> No. of Duplicate Rows in videos.csv: 0
 >> No. of Duplicate Rows in feeds.csv: 0
 >> No. of Duplicate Rows in watchedVideo.csv: 0


#### Notes on Data and Quality
- **4 csv files** are loaded as **4 different dataframes** using pandas module.
- Every file has different no. of rows and columns as shown above. 
- There are **NO** **missing values** or **duplicated entries** in the data to start with. This is a good starting point as we don't have to consider any assumption (E.g: to handle missing values etc.,) before starting with the data analysis.

P.S: 
- In the interest of time, for this challenge, we don't look into data quality issues at discrete feature level. 

***

## Descriptive Data Analysis
In this section, we perform descriptive analysis on the data being loaded into 4 different dataframes. Futher, our understanding about the data from descriptive analysis is documented in **Notes on Descriptive Analysis** sub-section. 

### Users data
We first look into the first few entries of the users data, identify datatypes of different features, and finally create a table of descriptive statistics on the features of the data. 

In [156]:
# Display first 5 rows in users data
df_users.head()

id           created_at        class  \
0  37e369a8-4477-4d5d-acd7-4e3b8059586e  2021-03-23 00:24:53      Creator   
1  27cf0bb6-3488-4662-916d-3ad04eeec9ca  2021-04-09 19:55:07       Viewer   
2  5ebd5e12-ba2a-4bf9-b7e4-92ad24711cd4  2021-04-08 21:48:45  Super Gamer   
3  30986592-0cbf-43bb-83ed-7501ab3bd373  2021-05-15 00:45:51      Creator   
4  25b49125-34d1-4694-a4a9-8415bcc3567b  2021-07-05 04:08:53       Viewer   

                 coordinates  
0    ('36.76775', '2.95924')  
1   ('43.71032', '-1.05366')  
2       ('51.26', '-2.1875')  
3  ('40.65538', '-74.38987')  
4  ('35.61452', '-88.81395')

In [157]:
# Datatypes of users data
df_users.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000 entries, 0 to 999
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   id           1000 non-null   object
 1   created_at   1000 non-null   object
 2   class        1000 non-null   object
 3   coordinates  1000 non-null   object
dtypes: object(4)
memory usage: 39.1+ KB


In [158]:
# Descriptive Statis on users data
df_users.describe(include='object')

id           created_at  class  \
count                                   1000                 1000   1000   
unique                                  1000                 1000      5   
top     37e369a8-4477-4d5d-acd7-4e3b8059586e  2021-03-23 00:24:53  Gamer   
freq                                       1                    1    213   

                  coordinates  
count                    1000  
unique                    626  
top     ('-6.25', '38.66667')  
freq                        6

### Videos data
In Videos data, we first look into the first few entries of the data, identify datatypes of different attributes, and finally create a table of descriptive statistics on the features of the data. 

In [159]:
# Display first 5 rows in videos data
df_videos.head()

id                          attribute_id  \
0  6e884071-8c0e-4782-aa40-e0fd78243360  4712e9e6-bef1-4646-b21e-ca5336901bf9   
1  ac90ff16-24b4-46bc-ae72-2da847f9e6c9  4f91d2b6-bf42-4d3f-a8e8-acc3b07f647e   
2  fb5eaf0d-e72c-4523-b5c5-4d9c323cef5f  80f5b187-f6e6-40f2-9b6e-fe62b67049a8   
3  bbf89dcb-7fbd-439d-9a7d-36ef7a1c2aa3  02b2b662-1b39-4d91-9417-dddc9451d8e1   
4  4c613ec6-fa6f-47e8-b4ab-658e2f6a644d  06c496ee-e884-42f9-ba39-010a66ed5296   

                               music_id                          animation_id  \
0  4ee0de5b-e27f-4d73-a3c6-52cf26e36c54  17dab81c-6180-4a4b-b539-91978a726cfd   
1  4ee0de5b-e27f-4d73-a3c6-52cf26e36c54  6d7a0545-0bca-4bf4-a306-931e57d8ffe2   
2  d53234f4-1efb-4299-97c1-11658118cba5  716c02df-bd3a-41c9-b2be-86fb74ff3255   
3  19e4db13-d854-4913-a8ca-3612ee1a6071  b7bd5204-2976-4eb0-ab71-508a848bc325   
4  b5b679a5-e769-4d95-82a6-f75e439b1ef9  17dab81c-6180-4a4b-b539-91978a726cfd   

   duration                               creator           created_at  \
0      5.34  73ec1ac2-d148-45ac-94a8-9bdff51730d7  2021-08-04 10:42:39   
1      3.88  0230f9af-e89f-4886-8322-52a38eaf991c  2021-11-27 10:06:43   
2      5.15  fbf93fbf-a935-436f-8516-8acb863e779f  2021-02-26 04:18:53   
3      6.50  f7a2adb8-44d7-4408-a2a9-3a593fc6d626  2021-11-27 02:08:02   
4      4.38  d9f9538f-1b7e-4f35-805d-924b309d75d8  2021-11-17 11:06:32   

   num_shares                               game_id  is_ar  
0          12  8bd4c7be-fe87-45ad-990d-d131cd3d9c77   True  
1           3  6ca0f645-a16b-42b0-9dc2-2ae952c83591   True  
2           4  6ca0f645-a16b-42b0-9dc2-2ae952c83591   True  
3           9  8bd4c7be-fe87-45ad-990d-d131cd3d9c77  False  
4          14  2609b6b5-3a08-451e-bce5-75fe506559d2  False

In [160]:
# Datatypes of videos data
df_videos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5000 entries, 0 to 4999
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            5000 non-null   object 
 1   attribute_id  5000 non-null   object 
 2   music_id      5000 non-null   object 
 3   animation_id  5000 non-null   object 
 4   duration      5000 non-null   float64
 5   creator       5000 non-null   object 
 6   created_at    5000 non-null   object 
 7   num_shares    5000 non-null   int64  
 8   game_id       5000 non-null   object 
 9   is_ar         5000 non-null   bool   
dtypes: bool(1), float64(1), int64(1), object(7)
memory usage: 395.5+ KB


In [161]:
# Descriptive Statis on videos data - Categorical
df_videos.describe(include='object')

id  \
count                                   5000   
unique                                  5000   
top     6e884071-8c0e-4782-aa40-e0fd78243360   
freq                                       1   

                                attribute_id  \
count                                   5000   
unique                                    11   
top     25e7fab9-abfb-4a74-a8cb-c3adbc6e3531   
freq                                     483   

                                    music_id  \
count                                   5000   
unique                                    10   
top     c3721b09-6896-465a-8e13-4d73efd06874   
freq                                     550   

                                animation_id  \
count                                   5000   
unique                                    15   
top     d167d0af-a662-48ad-bb39-5aa306fc8141   
freq                                     359   

                                     creator           created_at  \
count                                   5000                 5000   
unique                                   992                 5000   
top     ddfeab0f-274a-4cab-86c2-873cbd8e485e  2021-08-04 10:42:39   
freq                                      12                    1   

                                     game_id  
count                                   5000  
unique                                     5  
top     8bd4c7be-fe87-45ad-990d-d131cd3d9c77  
freq                                    1021

In [162]:
# Descriptive Statis on videos data - Numerical
df_videos.describe(exclude='object')

duration   num_shares is_ar
count   5000.000000  5000.000000  5000
unique          NaN          NaN     2
top             NaN          NaN  True
freq            NaN          NaN  2528
mean       4.536508     9.848800   NaN
std        1.402695     5.995526   NaN
min        2.100000     0.000000   NaN
25%        3.330000     5.000000   NaN
50%        4.555000    10.000000   NaN
75%        5.720000    15.000000   NaN
max        7.000000    20.000000   NaN

### Feeds data
In Feeds data too, we first look into the first few entries of the data, identify datatypes of different attributes, and finally create a table of descriptive statistics on the features of the data. 

In [163]:
# Display first 5 rows in feeds data
df_feeds.head()

id     name
0  27a87df4-da75-451f-b8f6-af9dd0a4ee00    music
1  42568c8a-efbe-4293-81a3-513e33094ade  fashion
2  6894af29-56cc-4b4a-8ee6-be9b3cb3e7f1   sports
3  c9d59470-c4ee-43ff-831d-b52508c8d848    games
4  582520b2-b30a-4ecb-b1b4-1806672fcfc4   travel

In [164]:
# Datatypes of feeds data
df_feeds.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10 entries, 0 to 9
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      10 non-null     object
 1   name    10 non-null     object
dtypes: object(2)
memory usage: 240.0+ bytes


In [165]:
# Descriptive Statis on feeds data
df_feeds.describe(include='object')

id   name
count                                     10     10
unique                                    10     10
top     27a87df4-da75-451f-b8f6-af9dd0a4ee00  music
freq                                       1      1

### WatchedVideos data
In watched videos data file, we first look into the first few entries of the data, identify datatypes of different attributes, and finally create a table of descriptive statistics on the features of the data. 

In [166]:
# Display first 5 rows in watched videos data
df_watched.head()

coordinates       os  duration  \
0  (Decimal('14.439490'), Decimal('105.291233'))      iOS      3.95   
1  (Decimal('40.866935'), Decimal('-73.845013'))      iOS      5.95   
2    (Decimal('48.666850'), Decimal('9.359456'))      iOS      0.69   
3   (Decimal('5.172365'), Decimal('-74.439009'))  Android      2.81   
4     (Decimal('4.063707'), Decimal('9.606159'))      iOS      5.88   

                                videoId                                userId  \
0  722e110c-2f81-46aa-9c25-4d26fd77519f  317a2bc3-5f95-4d35-9a94-1f2f38424c4f   
1  cc10d153-37f9-42b5-8d78-2365908acf2b  db031c3d-69e4-4c06-91fe-e9be0a3c173c   
2  7f3f7fce-f2ec-4dfd-bac8-92dc7865a365  8fb18b66-f548-42a7-b394-c1be963cac29   
3  8ffc47e3-32ce-4d24-a990-9baac556603b  960fb90e-eddc-478a-a64a-04b9dded3a05   
4  47e70477-7511-4bd0-801b-1651f951002f  1ce0c003-25e7-473f-8796-0fe04d6c275c   

  appVersion             dateTime                                feedId  
0      2.2.2  2021-08-04 09:39:07  30134b9e-1b79-43fb-b0c5-373af9237b2c  
1      2.2.1  2021-07-06 16:32:10  8078efe7-5bae-4528-9a66-b85904cc73ac  
2      2.2.0  2021-12-06 01:42:02  6894af29-56cc-4b4a-8ee6-be9b3cb3e7f1  
3      2.1.2  2021-12-04 21:02:50  42568c8a-efbe-4293-81a3-513e33094ade  
4      2.1.2  2021-11-28 16:12:39  582520b2-b30a-4ecb-b1b4-1806672fcfc4

In [167]:
# Datatypes of watched videos data
df_watched.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 49999
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   coordinates  50000 non-null  object 
 1   os           50000 non-null  object 
 2   duration     50000 non-null  float64
 3   videoId      50000 non-null  object 
 4   userId       50000 non-null  object 
 5   appVersion   50000 non-null  object 
 6   dateTime     50000 non-null  object 
 7   feedId       50000 non-null  object 
dtypes: float64(1), object(7)
memory usage: 3.4+ MB


### Notes on Descriptive analysis
- **users** data
    - No. of attributes = 4
    - All the attributes by default are object datatypes. In contrast, `created_at` should be **datetime** datatype and `coordinates` should be longitude, latitude float attributes. 
    - Unique identified `id` has 1000 unique values. This means, there are around 1000 unique users. 
- **videos** data
    - No. of attributes = 10
    - Here too, `created_at` should be **datetime** datatype. 
    - Unique identified `id` has 5000 unique values. This means, there are around 5000 unique videos. 
- **feeds** data
    - No. of attributes = 2
    - Unique identified `id` has 10 unique values. This means, there are around 10 unique feeds possible. 
- **watchedVideo** data
    - No. of attributes = 8
    - Here too, `coordinates` should be longitude, latitude 2 float attributes, and `dateTime` should be **datatime** datatype.
    - There are 50000 videos that were watched. This data might include videos that had multiple views. 

****

## Exploratory Data Analysis
In this section, we are going to understand different attributes by first making sure the datatypes as expected and then analysing datasets in either stadalone manner or by joining with two or more connected datasets.  

### Correcting datatypes and Feature engineering of user attributes

In [204]:
# 1. convert "created_at" to datetime 
df_users['created_at'] = pd.to_datetime(df_users['created_at'], format='%Y-%m-%d %H:%M:%S')

# 2.1 "coordinates" into 2 attributes (latitude-lat and longitude-lng)
df_users[["lat", "lng"]] = df_users["coordinates"].str.split(",", expand=True)
df_users['lat'] = df_users['lat'].str.replace("(",'', regex=True)
df_users['lng'] = df_users['lng'].str.replace(")",'', regex=True)

# 2.2 convert 2 new attributes to floats
df_users['lat'] = df_users['lat'].apply(lambda x: np.float(x.strip().replace("'",'')))
df_users['lng'] = df_users['lng'].apply(lambda x: np.float(x.strip().replace("'",'')))

#df_users['country'] = df_users.apply(lambda x: get_address_from_lat_lng(x.lat, x.lng), axis=1)

df_users.head()

id          created_at        class  \
0  37e369a8-4477-4d5d-acd7-4e3b8059586e 2021-03-23 00:24:53      Creator   
1  27cf0bb6-3488-4662-916d-3ad04eeec9ca 2021-04-09 19:55:07       Viewer   
2  5ebd5e12-ba2a-4bf9-b7e4-92ad24711cd4 2021-04-08 21:48:45  Super Gamer   
3  30986592-0cbf-43bb-83ed-7501ab3bd373 2021-05-15 00:45:51      Creator   
4  25b49125-34d1-4694-a4a9-8415bcc3567b 2021-07-05 04:08:53       Viewer   

                 coordinates       lat       lng  
0    ('36.76775', '2.95924')  36.76775   2.95924  
1   ('43.71032', '-1.05366')  43.71032  -1.05366  
2       ('51.26', '-2.1875')  51.26000  -2.18750  
3  ('40.65538', '-74.38987')  40.65538 -74.38987  
4  ('35.61452', '-88.81395')  35.61452 -88.81395

#### How many users are there by class?
Here, we calculate no. of user per class. That is, we get information such as how many users are creators, gamers, super creators, super gamers, or viewers. 

In [169]:
# Create dataframe for no. of users per class
df_users_class = df_users.groupby('class')['id'].count().reset_index()
df_users_class.columns = ['user_class','user_count']
df_users_class

user_class  user_count
0        Creator         186
1          Gamer         213
2  Super Creator         206
3    Super Gamer         209
4         Viewer         186

In [170]:
# Plot no. of users per class
df_users_class.plot_bokeh(kind='bar', title='No. of users by class',
                    x = 'user_class',
                    xlabel='Class of the User',
                    ylabel='No. of Users',
                    legend='top_left', alpha=0.6);

#### How many users are there by country or geographic location?
To get no. of users per country, we first have to convert user registered lat, lng into country attribute. This is done by using geopy's Nominatm library. The function `get_address_from_lat_lng` takes lat, lng as input and converts them into a address and returns a country. We limit geographical analysis to country level for this challenge and plot top 10 countries by users.   

In [220]:
# Create dataframe for no. of users per lat, lng
df_users_country = df_users.groupby(['lat','lng'])['id'].count().reset_index()
df_users_country['country'] = ''
df_users_country.head()

lat        lng  id country
0 -54.80000  -68.30000   2        
1 -44.39672  171.25364   2        
2 -38.65333  178.00417   2        
3 -38.16604  145.13643   2        
4 -37.90000  145.18333   2

Above takes indicates that users registered from around **626** unique locations. Now, we loop over above `df_users_country` dataframe to determine the country for every unique lat, lng. Since there are ~600 locations, converting all those lat, lng to country might take few minutes. To avoid accidently running this code again, we comment them after getting all countries for lat, lng and saving them in a seperate csv file. 

In [239]:
# Country from lat,lng
for i in range(df_users_country.shape[0]):
    lat = df_users_country.loc[i]['lat']
    lng = df_users_country.loc[i]['lng']
    country = get_address_from_lat_lng(lat, lng)
    #df_users_country.loc[i, 'country'] = country
    if (i % 25 == 0):
        print(i, lat, lng)#, country)

0 -54.8 -68.3 Argentina
25 -25.42944 -50.00639 Brasil
50 -11.94306 -76.70944 Perú
75 -3.74912 -73.25383 Perú
100 5.4709 100.24529 Malaysia
125 9.3337 122.8637 Philippines
150 13.48082 -86.58208 Nicaragua
175 18.57677 -72.22625 Ayiti
200 23.1959 86.51499 India
225 27.09978 -82.45426 United States
250 31.76212 -95.63079 United States
275 33.52253 -117.70755 United States
300 34.9 137.5 日本
325 36.61033 -88.31476 United States
350 37.65639 126.835 대한민국
375 39.96097 -75.60804 United States
400 40.93333 73.0 Кыргызстан
425 42.24113 -88.3162 United States
450 43.82634 144.09638 日本
475 45.49428 -122.86705 United States
500 48.06919 11.37703 Deutschland
525 50.56149 4.69889 België / Belgique / Belgien
550 51.30001 13.10984 Deutschland
575 52.66277 -2.01111 United Kingdom
600 54.90083 38.07083 Россия
625 68.79833 16.54165 Norge


In [240]:
# Save dataframe 
df_users_country.to_csv('lat_lng_country.csv')

In [282]:
# Top 10 countries by users
df_users_country_all = df_users_country.groupby(['country'])['id'].count().reset_index()
df_users_country_all.columns = ['country','user_count']
df_users_country_top10 = df_users_country_all.sort_values(by='user_count', ascending=False).head(10)
print(df_users_country_all.shape[0])
df_users_country_top10

109


country  user_count
70    United States         103
28            India          45
80           Россия          34
105              中国          31
9            Brasil          31
20      Deutschland          28
69   United Kingdom          26
26           France          22
30           Italia          18
29        Indonesia          13

In [281]:
# User Count by lat, lng -> for mapplot
df_users_count_lat_lng = df_users_country.groupby(['lat','lng'])['id'].count().reset_index()
df_users_count_lat_lng

lat        lng  id
0   -54.80000  -68.30000   1
1   -44.39672  171.25364   1
2   -38.65333  178.00417   1
3   -38.16604  145.13643   1
4   -37.90000  145.18333   1
..        ...        ...  ..
621  60.97944   76.92421   1
622  61.72744   17.10558   1
623  64.56350   39.83020   1
624  65.84811   24.14662   1
625  68.79833   16.54165   1

[626 rows x 3 columns]

In [295]:
# Pie chart of Top10 countries by users
df_users_country_top10.plot_bokeh(kind='pie', x="country", y="user_count", 
                                  title="Top 10 countries by user base",
                                  colormap=["blue", "red", "yellow", "green", "purple", "orange", "grey",'pink','brown','lightblue']
                                 );

__x__values_original


In [307]:
# Plot No. of Users by country using Folium library 
m = folium.Map(tiles="StamenToner", min_zoom=2)

# No. of users and lat, lng
nusers = df_users_count_lat_lng['id'].astype(float)
lat = df_users_count_lat_lng['lat'].astype(float)
lng = df_users_count_lat_lng['lng'].astype(float)

# Add nusers data as Heatmap 
m.add_child(HeatMap(zip(lat,lng, nusers),radius=0))

# Add dot per user on top of Heatmap 
df_users_count_lat_lng.apply(plot_dots_folium, axis=1)
m.fit_bounds(m.get_bounds())
m

In [337]:
# Plot Folium map 
m = folium.Map(zoom_start=11)

# Add additional Tiles
folium.TileLayer('Stamen Terrain').add_to(m)
folium.TileLayer('Stamen Toner').add_to(m)
folium.TileLayer('Stamen Water Color').add_to(m)
folium.TileLayer('cartodbpositron').add_to(m)

# Add marker cluster
marker_cluster = MarkerCluster(name='No. of users clustered icons',
                               overlay=True, control=False, icon_create_function=None)

for k in range(len(lat)):
    location = lat[k], lng[k]
    marker = folium.Marker(location=location)
    popup = 'lng:{}<br> lat:{}<br> user_count:{}'.format(location[1], location[0], nusers[k])
    folium.Popup(popup).add_to(marker)
    marker_cluster.add_child(marker)

marker_cluster.add_to(m)

folium.LayerControl().add_to(m);
m

### How many users have registered over date and time?
To answer this question, we can make use of the `created_at` attribute to calculate the no. of users over date and time. 
1. First, we compute no. of users registered every month during 2021.
2. Next, we find no. of users registered by day of week
3. Finally, we compute no. of users registered by hour of day.

#### Table of no. of users by month

In [354]:
# No. of users registered per month in 2021
df_users_month = df_users.groupby(df_users['created_at'].dt.month)['id'].count().reset_index()
df_users_month.columns = ['Month', 'user_count']

# Convert month number to Month name
df_users_month['Month'] = df_users_month['Month'].apply(lambda x: calendar.month_abbr[x])
df_users_month

Month  user_count
0    Jan          90
1    Feb          79
2    Mar          83
3    Apr          85
4    May          77
5    Jun          83
6    Jul         103
7    Aug          88
8    Sep          93
9    Oct          82
10   Nov          95
11   Dec          42

#### Plot of no. of users by month

In [361]:
# Line of no. of users by month in 2021
df_users_month.plot_bokeh(kind='line', figsize=(800, 450), title="Monthly user registrations in 2021",
              xlabel="Months in 2021", ylabel="No. of users",
              x='Month', y='user_count',
              ylim=(0, 150),
              plot_data_points=True, plot_data_points_size=10, marker="o");

#### Table of no. of users by Day of the week

In [364]:
# No. of users registered by day of week
df_users_dow = df_users.groupby(df_users['created_at'].dt.dayofweek)['id'].count().reset_index()
df_users_dow.columns = ['day_of_week', 'user_count']

# Convert DOW number to DOW name
df_users_dow['day_of_week'] = df_users_dow['day_of_week'].apply(lambda x: calendar.day_abbr[x])
df_users_dow

day_of_week  user_count
0         Mon         150
1         Tue         136
2         Wed         146
3         Thu         147
4         Fri         155
5         Sat         145
6         Sun         121

#### Plot of no. of users by Day of the week

In [366]:
# Line of no. of users by Day of week
df_users_dow.plot_bokeh(kind='line', figsize=(800, 450), title="Days of the week user registrations in 2021",
              xlabel="Days of the Week", ylabel="No. of users",
              x='day_of_week', y='user_count',
              ylim=(0, 200),
              plot_data_points=True, plot_data_points_size=10, marker="o");

#### Table of no. of users by Hour of the day

In [450]:
# No. of users registered by hour of the day
df_users_hr = df_users.groupby(df_users['created_at'].dt.hour)['id'].count().reset_index()
df_users_hr.columns = ['hour_of_day', 'user_count']

# Create hour category
df_users_hr['hr_category'] = pd.cut(df_users_hr['hour_of_day']+1, np.arange(0, 27, 3))

# create users_hr_category
df_users_hr_cat = df_users_hr.groupby('hr_category')['user_count'].sum().reset_index()
df_users_hr_cat['hr_category'] = df_users_hr_cat['hr_category'].apply(lambda x: str(x))
df_users_hr_cat['user_count_pct'] = df_users_hr_cat['user_count'] / df_users_hr_cat['user_count'].sum() * 100
df_users_hr_cat

hr_category  user_count  user_count_pct
0      (0, 3]         120            12.0
1      (3, 6]         116            11.6
2      (6, 9]         102            10.2
3     (9, 12]         149            14.9
4    (12, 15]         126            12.6
5    (15, 18]         130            13.0
6    (18, 21]         132            13.2
7    (21, 24]         125            12.5

#### Plot of no. of users by Hour of the day

In [453]:
# Plot no. of users by hour of the day 
df_users_hr_cat.plot_bokeh(kind='pie', x="hr_category", y='user_count_pct', title="Percent of users by hour of the day",
                       colormap=["blue", "red", "yellow", "green", "purple", "orange", "grey", 'pink'],
                       line_color="grey");

__x__values_original


### Correcting datatypes in videos dataset

***